File: src/com/suushiemaniac/cubing/bld/optim/AlgComparator.java
package com.suushiemaniac.cubing.bld.optim;

import com.suushiemaniac.cubing.alglib.alg.Algorithm;

import java.util.Comparator;

public class AlgComparator implements Comparator<Algorithm> {
	protected static AlgComparator INST;

	public static AlgComparator INST() {
		if (INST == null) {
			INST = new AlgComparator();
		}

		return INST;
	}

	public static float scoreAlg(Algorithm alg) {
		return INST().score(alg);
	}

	protected AlgComparator() {}

	@Override
	public int compare(Algorithm alg, Algorithm otherAlg) {
		return Float.compare(this.score(otherAlg), this.score(alg));
	}

	public float score(Algorithm alg) {
		return (
			2 * this.lengthScore(alg)
			+ this.rotationScore(alg)
			+ this.subGroupScore(alg)
		) / 4f;
	}

	protected int lengthScore(Algorithm alg) {
		return -2 * alg.moveLength() + 26;
	}

	protected int rotationScore(Algorithm alg) {
		return -5 * alg.getRotationGroup().size() + 10;
	}

	protected int subGroupScore(Algorithm alg) {
		return -5 * alg.getSubGroup().size() + 25;
	}
}

File: src/com/suushiemaniac/cubing/bld/optim/AlgComparator.kt
package com.suushiemaniac.cubing.bld.optim

import com.suushiemaniac.cubing.alglib.alg.Algorithm

class AlgComparator protected constructor() : Comparator<Algorithm> {
    override fun compare(alg: Algorithm, otherAlg: Algorithm): Int {
        return java.lang.Float.compare(this.score(otherAlg), this.score(alg))
    }

    fun score(alg: Algorithm): Float {
        return (2 * this.lengthScore(alg)
                + this.rotationScore(alg)
                + this.subGroupScore(alg)) / 4f
    }

    protected fun lengthScore(alg: Algorithm): Int {
        return -2 * alg.moveLength() + 26
    }

    protected fun rotationScore(alg: Algorithm): Int {
        return -5 * alg.rotationGroup.size() + 10
    }

    protected fun subGroupScore(alg: Algorithm): Int {
        return -5 * alg.subGroup.size() + 25
    }

    companion object {
        val SINGLETON = AlgComparator()

        fun scoreAlg(alg: Algorithm): Float {
            return SINGLETON.score(alg)
        }
    }
}